airbyte-integrations/connectors/source-zuora/source_zuora/source.py

```diff
@@ -0,0 +1,427 @@
+
+        return [json_schema]
+
+
+class ZuoraSubmitJob(ZuoraStream):
+
+    http_method = "POST"
+
+    def __init__(self, query: str, config: Dict):
+        super().__init__(config)
+        self.query = query
+
+    def path(self, **kwargs) -> str:
+        return "/query/jobs"
+
+    def request_body_json(self, **kwargs) -> Optional[Mapping]:
+        params = self._request_params(stream_state=None)
+        params["query"] = self.query
+        return params
+
+    def parse_response(self, response: requests.Response, **kwargs) -> List[str]:
+
+        """
+        Response example:
+        {'data':
+            {
+                'id': 'c6f25f91-5357-4fec-a00d-9009cc1ae856',
+                'query': 'DESCRIBE account',
+                'useIndexJoin': False,
+                'sourceData': 'LIVE',
+                'queryStatus': 'accepted',
+                'remainingRetries': 3,
+                'retries': 3,
+                'updatedOn': '2021-07-26T15:33:48.287Z',
+                'createdBy': '84f78cea-8a5b-4332-933f-27439fe3b87b'
+            }
+        }
+        """
+        return [response.json()["data"]["id"]]
+
+
+class ZuoraJobStatusCheck(ZuoraStream):
+    def __init__(self, job_id: str, config: Dict):
+        super().__init__(config)
+        self.job_id = job_id
+
+    def path(self, **kwargs) -> str:
+        return f"/query/jobs/{self.job_id}"
+
+    def parse_response(self, response: requests.Response, **kwargs) -> List[str]:
+        return [response.json()["data"]["dataFile"]]
+
+    def _send_request(self, request: requests.PreparedRequest, request_kwargs: Mapping[str, Any]) -> requests.Response:
+
+        """
+        {'data':
+            {
+                'id': 'c6f25f91-5357-4fec-a00d-9009cc1ae856',
+                'query': 'DESCRIBE account',
+                'useIndexJoin': False,
+                'sourceData': 'LIVE',
+                'queryStatus': 'completed',
+                'dataFile': 'https://owl-auw2-sbx01-query-result.s3.us-west-2.amazonaws.com/c6f25f91-5357-4fec-a00d-9009cc1ae856_2779514650704989.jsonl?....',
+                'outputRows': 53,
+                'processingTime': 516,
+                'remainingRetries': 3,
+                'retries': 3,
+                'updatedOn': '2021-07-26T15:33:48.803Z',
+                'createdBy': '84f78cea-8a5b-4332-933f-27439fe3b87b'
+            }
+        }
+        """
+
+        # Define the job error statuses
+        errors = ["failed", "canceled", "aborted"]
+        # Error msg: the cursor_field cannot be resolved
+        cursor_error = "Column 'updateddate' cannot be resolved"
+        # Error msg: cannot process object
+        obj_read_error = "failed to process object"
+
+        status = None
+        success = "completed"
+        while status != success:
+            response: requests.Response = self._session.send(request, **request_kwargs)
+            job_check = response.json()
+            status = job_check["data"]["queryStatus"]
```

Comment:
is there an opportunity of infinity loop?
Please add a comment about this condition.

Reply:
There is no opportunity for the infinity loop because the operation performs on the server-side, there are query run-time limitations: if the query time is longer than 120 min, the server will output the error with the corresponding message for the user in the output, by raising `ZOQLQueryFailed` exception. I'll add this comment to the code.